internal/ingress/controller/status.go

```diff
@@ -93,19 +92,25 @@ func setupLeaderElection(config *leaderElectionConfig) {
 		Host:      hostname,
 	})
 
-	lock := resourcelock.ConfigMapLock{
-		ConfigMapMeta: metav1.ObjectMeta{Namespace: k8s.IngressPodDetails.Namespace, Name: config.ElectionID},
-		Client:        config.Client.CoreV1(),
-		LockConfig: resourcelock.ResourceLockConfig{
+	//TODO: Switch configmapsleases to leases in the next version.
```

Comment:
Would like to ask for a soft transition here, let me explain:
* Test if we CAN use Leases, then we use it.
* If not, test if we can use ConfigMapLeases, then we use it
* Finally, if everything goes wrong, fallback to ConfigMapLock

This way, users that already rely on ConfigMap wont need to change/break their environment. Forcefully changing to `Leases` may break environments where the admins just update the image, and not the Roles, etc.

Makes sense?

Reply:
I did not get completely what you are suggesting. First, leaderelection should try to acquire resourcelock using `leases`,  and if it fails, try `configmapleases`, and then `configmaps`?
Please explain.

Reply:
TBH, admins not updating the Roles does not feel like a reason enough to me. We will be mentioning that in the release notes, so admins should be noticing that and acting accordingly.
And anyways if we delay this too, after we revendor K8s, we have to get rid of the old resourcelocks.
WDYT?

Reply:
@shafeeqes this is true in theory, but not in practice and we must be empathetic with the admins on this :)

As a previous admin, I know that sometimes in a rush we simply update the image. This is a breaking change from the users perspective. If someone just update the image on a cluster previously created with old manifests (instead of reapplying all the manifests, and well, `kubectl set image` is here for that!) Ingress wont start, and this may lead into their ingresses being unavailable.

So, my suggestion is:
- Test if ingress can do a get/update/list/watch on leases. If so, let's use it
- Otherwise, fallback to the previously working scenario.

Reply:
What about exposing a CLI flag similar to `--election-id` for configuring the lock type? This would prevent this fake call to check whether the nginx-ingress pod has required privilege and it would also decouple the leader election lock migration from nginx-ingress releases. Users of nginx-ingress can configure this as they want when they are ready (i.e., when they have provided the required RBAC privileges).